src/main/java/games/strategy/io/IoUtils.java

```diff
@@ -0,0 +1,120 @@
+    });
+  }
+
+  /**
+   * Applies the specified function to an in-memory input stream that wraps the specified byte array and returns the
+   * function result.
+   *
+   * @param bytes The byte array from which to read.
+   * @param function The function to apply to an input stream containing {@code bytes}.
+   *
+   * @return The function result.
+   *
+   * @throws IOException If {@code function} encounters an error while reading from the input stream.
+   */
+  public static <T> T readFromMemory(final byte[] bytes, final InputStreamFunction<T> function) throws IOException {
+    checkNotNull(bytes);
+    checkNotNull(function);
+
+    // NB: ByteArrayInputStream does not need to be closed
+    return function.apply(new ByteArrayInputStream(bytes));
+  }
+
+  /**
+   * Invokes the specified consumer with an in-memory output stream and returns the bytes written.
+   *
+   * @param consumer The consumer whose output will be captured and returned.
+   *
+   * @return The bytes written by the consumer to the output stream.
+   *
+   * @throws IOException If {@code consumer} encounters an error while writing to the output stream.
+   */
+  public static byte[] writeToMemory(final OutputStreamConsumer consumer) throws IOException {
+    checkNotNull(consumer);
+
+    // NB: ByteArrayOutputStream does not need to be closed
+    final ByteArrayOutputStream os = new ByteArrayOutputStream();
+    consumer.accept(os);
+    return os.toByteArray();
+  }
+
+  /**
+   * An operation that accepts an {@link InputStream} and returns no result.
+   */
+  @FunctionalInterface
+  public interface InputStreamConsumer {
+    /**
+     * Performs the operation using the specified input stream.
+     *
+     * @param is The input stream from which the consumer will read.
+     *
+     * @throws IOException If an I/O error occurs while reading from the input stream.
+     */
+    void accept(InputStream is) throws IOException;
+  }
+
+  /**
+   * A function that accepts an {@link InputStream} and produces a result.
+   *
+   * @param <R> The type of the function result.
+   */
+  @FunctionalInterface
+  public interface InputStreamFunction<R> {
+    /**
+     * Applies this function to the specified input stream.
+     *
+     * @param is The input stream from which the function will read.
+     *
+     * @return The function result.
+     *
+     * @throws IOException If an I/O error occurs while reading from the input stream.
+     */
+    R apply(InputStream is) throws IOException;
+  }
+
+  /**
+   * An operation that accepts an {@link OutputStream} and returns no result.
+   */
+  @FunctionalInterface
+  public interface OutputStreamConsumer {
+    /**
+     * Performs the operation using the specified output stream.
+     *
+     * @param os The output stream to which the consumer will write.
+     *
+     * @throws IOException If an I/O error occurs while writing to the output stream.
+     */
+    void accept(OutputStream os) throws IOException;
+  }
```

Comment:
I believe you could extend Function and Consumer and override the default methods.
This way the method naming is enforced to be consistent

Reply:
Unfortunately, that won't work. 😞   The `Function` and `Consumer` methods are declared to not throw a checked exception, and, due to the LSP, subtypes cannot widen the exception specification.  I extracted these interfaces specifically to allow implementors to continue to throw `IOException` without having to wrap it in an unchecked exception.

Reply:
I wasn't sure if exceptions could be changed when overriding a method, but it makes sense, because if we were to cast it to it's supertype this information would be lost and no try catch block would be required although this method explicitly throws an exception.

Reply:
Exactly.

Note that you _can_ change the exception specification as long as it is only _narrowed_.  For example, if the interface method was defined as

```java
void accept(InputStream is) throws IOException;
```

Then, if the implementation happens to never throw `IOException`, it is free to declare the method as

```java
@Override
public void accept(InputStream is) {
  // . . .
}
```